Spigot-Server-Patches/0139-Implement-Profile-api.patch

```diff
@@ -0,0 +1,674 @@
++            return null;
++        }
++    }
++}
+\ No newline at end of file
+diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
+index 767c384..241306e 100644
+--- a/src/main/java/net/minecraft/server/EntityHuman.java
++++ b/src/main/java/net/minecraft/server/EntityHuman.java
+@@ -22,6 +22,10 @@ import org.bukkit.event.player.PlayerDropItemEvent;
+ import org.bukkit.event.player.PlayerVelocityEvent;
+ import org.bukkit.util.Vector;
+ // CraftBukkit end
++// PaperSpigot start
++import com.destroystokyo.paper.profile.PlayerProfile;
++import com.destroystokyo.paper.profile.ProfileUtils;
++// PaperSpigot end
+ 
+ public abstract class EntityHuman extends EntityLiving {
+ 
+@@ -1205,6 +1209,14 @@ public abstract class EntityHuman extends EntityLiving {
+         return this.bR;
+     }
+ 
++    // PaperSpigot start - bukkit profile method
++    private final PlayerProfile bukkitProfile = ProfileUtils.toPaperWithProperties(getProfile());
++
++    public PlayerProfile getBukkitProfile() {
++        return bukkitProfile;
++    }
++    // PaperSpigot end
++
+     public EntityHuman.EnumBedResult a(BlockPosition blockposition) {
+         if (!this.world.isClientSide) {
+             if (this.isSleeping() || !this.isAlive()) {
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+index 4521786..94125fe 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+@@ -21,6 +21,10 @@ import org.bukkit.configuration.serialization.SerializableAs;
+ import org.bukkit.entity.Player;
+ import org.bukkit.metadata.MetadataValue;
+ import org.bukkit.plugin.Plugin;
++// PaperSpigot start
++import com.destroystokyo.paper.profile.ProfileUtils;
++import com.destroystokyo.paper.profile.PlayerProfile;
++// PaperSpigot end
+ 
+ @SerializableAs("Player")
+ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializable {
+@@ -32,11 +36,25 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
+         this.server = server;
+         this.profile = profile;
+         this.storage = (WorldNBTStorage) (server.console.worlds.get(0).getDataManager());
++        // PaperSpigot start - store our profile
++        this.paperProfile = ProfileUtils.toPaper(profile);
++    }
++    private final PlayerProfile paperProfile;
++
++    protected CraftOfflinePlayer(CraftServer server, PlayerProfile profile) {
++        this.server = server;
++        this.profile = ProfileUtils.toMojang(profile);
++        this.storage = (WorldNBTStorage) (server.console.worlds.get(0).getDataManager());
++        this.paperProfile = profile;
++    }
+ 
++    public PlayerProfile getProfile() {
++        return paperProfile;
+     }
+ 
+-    public GameProfile getProfile() {
++    public GameProfile getNMSProfile() {
```

Comment:
Bad idea renaming this - change the API method to `getPlayerProfile` or something.